### PR TITLE
Add options to close (spec)

### LINKED ISSUE #636: Add options to close
per 2024-02-13, add options to close markup in syntax

### PR DESCRIPTION
**Pull request by mihnita**
Closes #636

### CODE REVIEW

File: spec/syntax.md [hunk 1]
```diff
@@ -726,13 +730,13 @@ attribute = "@" identifier [[s] "=" [s] (literal / variable)]
...
 > ```
-> In French, "{|bonjour| @translate=no}" is a greeting
+> In French, "{|bonjour| @translate=no/}" is a greeting
 > ```
 >
 > A _message_ with _markup_ that should not be copied:
 >
 > ```
-> Have a {+span @can-copy}great and wonderful{-span @can-copy} birthday!
+> Have a {#span @can-copy}great and wonderful{/span @can-copy} birthday!
```
Comment (aphillips) [inline spec/syntax.md:739]: Good catch!!!

**Review by aphillips — APPROVED** [review]
LGTM

There might be a quibble about whether we want separate open/standalone/close productions or separate `/` from `}` on standalone, but I think this is pretty darned clean.

Thanks!!

**Comment by mihnita** [comment]
> There might be a quibble about whether we want separate open/standalone/close productions or separate `/` from `}` on standalone, but I think this is pretty darned clean.

ACK.

I thought about this, and decided to use this style to bring it closer to the json data model, where we have 3 completely separated structures for markup-open/standalone/close.

They are also not unified in the data-model/README:
```
type Markup = MarkupOpen | MarkupStandalone | MarkupClose;
interface MarkupOpen {
    type: "markup";
    kind: "open";
    ...
}
interface MarkupStandalone {
    type: "markup";
    kind: "standalone";
    ...
}
interface MarkupClose {
    type: "markup";
     kind: "close";
    ....
}
```

On the other side, in the data-model dtd they are all unified.

---

I would prefer the unified style.

So the readme would be more like

```
type Markup = MarkupOpen | MarkupStandalone | MarkupClose;
interface Markup {
    type: "markup";
    kind; // one of "open", "close", "standalone"
    ...
}
```

But I didn't want to include anything controversial.

**Comment by mihnita** [comment]
I would also like to fix this issue: "Extra spaces in markup #650" (that I've just created)

File: spec/data-model/README.md [hunk 2]
```diff
@@ -266,6 +266,7 @@ interface MarkupClose {
   type: "markup";
   kind: "close";
   name: string;
+  options?: Option[];
```
Comment (eemeli) [inline spec/data-model/README.md:269]: One of the presented reasons for including options on close is that this removes the differences between closing markup and other parts. So we can simplify all of the constructs here into:
```ts
interface Markup {
  type: "markup";
  kind: "open" | "standalone" | "close";
  name: string;
  options: Option[];
  attributes: Attribute[];
}
```

File: spec/data-model/message.json [hunk 3]
```diff
@@ -157,6 +157,7 @@
         "type": { "const": "markup" },
         "kind": { "const": "close" },
         "name": { "type": "string" },
+        "options": { "$ref": "#/$defs/options" },
```
Comment (eemeli) [inline spec/data-model/message.json:160]: Same change here as for the TS; we don't need to have separate markup-open, markup-standalone, and markup-close definitions any more.

File: spec/message.abnf [hunk 4]
```diff
@@ -32,11 +32,9 @@ annotation            = function
                       / private-use-annotation
                       / reserved-annotation
 
-; Markup; Note that standalone markup is part of the first line
-markup       = "{" [s] markup-open *(s attribute) [s] ["/"] "}"
-             / "{" [s] markup-close *(s attribute) [s] "}"
-markup-open  = "#" identifier *(s option)
-markup-close = "/" identifier
+markup = "{" [s] "#" identifier *(s option) *(s attribute) [s] "}"  ; open
+       / "{" [s] "#" identifier *(s option) *(s attribute) [s] "/}" ; standalone
```
Comment (eemeli) [inline spec/message.abnf:36]:
```suggestion
markup = "{" [s] "#" identifier *(s option) *(s attribute) [s] ["/"] "}"  ; open and standalone
```

File: spec/syntax.md [hunk 5]
```diff
@@ -678,21 +678,25 @@ It MAY include _options_.
...
 is a _pattern_ part ending a span.
-Unlike the other forms, it does not include _options_.
 
 ```abnf
-markup       = "{" [s] markup-open *(s attribute) [s] ["/"] "}"
-             / "{" [s] markup-close *(s attribute) [s] "}"
-markup-open  = "#" identifier *(s option)
-markup-close = "/" identifier
+markup = "{" [s] "#" identifier *(s option) *(s attribute) [s] "}"  ; open
+       / "{" [s] "#" identifier *(s option) *(s attribute) [s] "/}" ; standalone
```
Comment (eemeli) [inline spec/syntax.md:684]:
```suggestion
markup = "{" [s] "#" identifier *(s option) *(s attribute) [s] ["/"] "}"  ; open and standalone
```

File: spec/syntax.md [hunk 6]
```diff
@@ -726,13 +730,13 @@ attribute = "@" identifier [[s] "=" [s] (literal / variable)]
 > A _message_ including a _literal_ that should not be translated:
 >
 > ```
-> In French, "{|bonjour| @translate=no}" is a greeting
+> In French, "{|bonjour| @translate=no/}" is a greeting
```
Comment (eemeli) [inline spec/syntax.md:733]: This is a literal, not markup.

```suggestion
> In French, "{|bonjour| @translate=no}" is a greeting
```

**Review by eemeli — CHANGES_REQUESTED** [review]
Overall ok, but see inline simplifications.

File: spec/data-model/README.md (hunk 2 above)
Comment (stasm) [reply spec/data-model/README.md:269]: Sounds like @mihnita would prefer this form as well: https://github.com/unicode-org/message-format-wg/pull/649#issuecomment-1945282227.

File: spec/message.abnf (hunk 4 above)
Comment (stasm) [reply spec/message.abnf:36]: There's precedent in our ABNF of refactoring productions to make them more readable for humans at the cost of increasing the lookahead. However here keeping `open` and `standalone` as separate alternatives results in `LL(infinity)` due to `*` on options and attributes. I'd prefer to not go that far, so I agree with @eemeli's suggestion.

File: spec/data-model/README.md (hunk 2 above)
Comment (mihnita) [reply spec/data-model/README.md:250]: Done

File: spec/data-model/message.json (hunk 3 above)
Comment (mihnita) [reply spec/data-model/message.json:138]: Done

**Comment by mihnita** [comment]
All feedback implemented, thank you all!
(ready for round two :-)

File: spec/data-model/README.md [hunk 7]
```diff
@@ -236,35 +236,16 @@ interface UnsupportedAnnotation {
 
 ## Markup
 
-A `Markup` object is either `MarkupOpen`, `MarkupStandalone`, or `MarkupClose`,
-which are differentiated by `kind`.
+A `Markup` object is either can be _open_, _standalone_, or _close_,
+and that information is stored in `kind`.
```
Comment (eemeli) [inline spec/data-model/README.md:240]:
```suggestion
A `Markup` object has a `kind` of either `"open"`, `"standalone"`, or `"close"`,
each corresponding to _open_, _standalone_, and _close_ _markup_.
```

**Review by eemeli — APPROVED** [review]

**Pull request merged**